File: leetcode/2391/Solution.py
from typing import List

class Solution:
    M_CHAR = 'M'
    P_CHAR = 'P'
    G_CHAR = 'G'
    def garbageCollection(self, garbage: List[str], travel: List[int]) -> int:
        ans = len(garbage[0])
        currentTravel = 0
        mTravel = 0
        pTravel = 0
        gTravel = 0
        for i in range(1, len(garbage)):
            ans += len(garbage[i])
            currentTravel += travel[i - 1]
            for cur in garbage[i]:
                if cur == Solution.M_CHAR:
                    mTravel = currentTravel
                if cur == Solution.P_CHAR:
                    pTravel = currentTravel
                if cur == Solution.G_CHAR:
                    gTravel = currentTravel
        return ans + mTravel + pTravel + gTravel

if __name__ == '__main__':
    sol = Solution()
    garbage = ["G","P","GP","GG"]
    travel = [2,4,3]
    print("test:", sol.garbageCollection(garbage, travel))

File: leetcode/2391/Solution.java
class Solution {
    private static final char M = 'M';
    private static final char P = 'P';
    private static final char G = 'G';
    public int garbageCollection(String[] garbage, int[] travel) {
        int n = garbage.length;
        int ans = garbage[0].length();
        int mTravel = 0;
        int pTravel = 0;
        int gTravel = 0;
        int currentTravel = 0;
        for (int i = 1; i < n; i++) {
            ans += garbage[i].length();
            currentTravel += travel[i - 1];
            for (char type : garbage[i].toCharArray()) {
                if (type == M) mTravel = currentTravel;
                if (type == P) pTravel = currentTravel;
                if (type == G) gTravel = currentTravel;
            }
        }
        return ans + mTravel + pTravel + gTravel;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        String[] garbage = {"G","P","GP","GG"};
        int[] travel = {2, 4, 3};
        System.out.println("test: " + sol.garbageCollection(garbage, travel));
    }
}